rabbit-openqa.py

```diff
@@ -0,0 +1,209 @@
+                                    self.map_openqa_result(result),
+                                    test + '@' + machine)
+        url = self.api.makeurl(['status_reports', 'published', staging, 'images', 'reports', buildid])
+        http_POST(url, data=xml)
+    
+    def openqa_create(self, iso, test, machine, id):
+        print('openqa_create', iso, test, machine, id)
+
+    def openqa_check_xml(self, id, state, name):
+        check = ET.Element('check')
+        se = ET.SubElement(check, 'url')
+        se.text = "https://openqa.suse.de/tests/{}".format(id)
+        se = ET.SubElement(check, 'state')
+        se.text = state
+        se = ET.SubElement(check, 'name')
+        se.text = name
+        return ET.tostring(check)
+
+class Listener(object):
+    def __init__(self, amqp_prefix, amqp_url):
+        self.projects = []
+        self.amqp_prefix = amqp_prefix
+        self.amqp_url = amqp_url
+        connection = pika.BlockingConnection(pika.URLParameters(amqp_url))
+        self.channel = connection.channel()
+
+        self.channel.exchange_declare(exchange='pubsub', exchange_type='topic', passive=True, durable=True)
+
+        result = self.channel.queue_declare(exclusive=True)
+        queue_name = result.method.queue
+
+        self.channel.queue_bind(exchange='pubsub',
+                                queue=queue_name,routing_key='#')
+        self.channel.basic_consume(self.on_message,
+                                   queue=queue_name,
+                                no_ack=True)
+
+        print(' [*] Waiting for logs. To exit press CTRL+C')
+
+    def add(self, project):
+        self.projects.append(project)
+
+    def on_published_repo(self, payload):
+        for p in self.projects:
+            p.check_published_repo(str(payload['project']), str(payload['repo']), str(payload['buildid']))
+
+    def on_openqa_create(self, payload):
+        for p in self.projects:
+            p.openqa_create(str(payload.get('ISO', '')), str(payload['TEST']), str(payload['MACHINE']), str(payload['id']))
+
+    def on_openqa_restart(self, payload):
+        print(payload)
+
+    def on_openqa_done(self, payload):
+        for p in self.projects:
+            p.openqa_done(str(payload.get('ISO', '')), payload['TEST'], payload['MACHINE'], payload['id'], payload['result'])
+
+    def on_message(self, unused_channel, method, properties, body):
+        if method.routing_key == '{}.obs.repo.published'.format(amqp_prefix):
+            self.on_published_repo(json.loads(body))
+
+        if method.routing_key == '{}.openqa.job.done'.format(amqp_prefix):
+            self.on_openqa_done(json.loads(body))
+        if method.routing_key == '{}.openqa.job.create'.format(amqp_prefix):
+            self.on_openqa_create(json.loads(body))
+        if method.routing_key == '{}.openqa.job.restart'.format(amqp_prefix):
+            self.on_openqa_restart(json.loads(body))
+
+    def listen(self):
+        self.channel.start_consuming()
```

Comment:
Curious, how does this handle disconnects? When experimenting with a rabbit concept last week the connection would periodically close. Would imagine graceful reconnects would be important.

Reply:
This is basically the basic rabbitmq tutorial code - but you can have callbacks for on_disconnect also and reconnect. 

For not getting disconnects, it's rather important to define a heartbeat_interval. And also we would need to define a server side message queue instead of using pubsub directly.

Reply:
https://pika.readthedocs.io/en/0.9.14/examples/asynchronous_consumer_example.html is most likely the better base.